core/backend/src/StashManager.ts

```diff
@@ -0,0 +1,458 @@
+  readonly timestamp: string;
+
+  /** Description of the stash */
+  readonly description: string;
+
+  /** Hash of the stash */
+  readonly hash: string;
+
+  /** Parent changeset of the stash */
+  readonly parentChangeset: ChangesetIdWithIndex;
+
+  /** ID sequences for the stash */
+  readonly idSequences: {
+    element: Id64String;
+    instance: Id64String;
+  };
+
+  /** Transaction properties for the stash */
+  readonly txns: TxnProps[];
+
+  /** Number of locks acquired by the stash */
+  readonly acquiredLocks: number;
+}
+
+/**
+ * Properties for creating a stash
+ * @internal
+ */
+export interface CreateStashProps {
+  /** Briefcase database instance */
+  readonly db: BriefcaseDb;
+  /** description of the stash */
+  readonly description: string;
+  /** discard all local changes and unless keepLocks flag is set, all locks will be released */
+  readonly discardLocalChanges?: true;
+  /** keep all locks after discarding local changes */
+  readonly retainLocks?: true;
+}
+
+/**
+ * Arguments for stash operations
+ * @internal
+ */
+export interface StashArgs {
+  readonly db: BriefcaseDb;
+  readonly stash: Id64String | StashProps;
+}
+
+/**
+ * Arguments for applying or restoring a stash.
+ *
+ * Extends {@link StashArgs} with an additional `method` property to specify the operation.
+ *
+ * @property method - Specifies the stash operation to perform.
+ *   - `"apply"`: Apply the stash without removing it.
+ *   - `"restore"`: Apply the stash and remove it from the stash list.
+ * @internal
+ */
+export interface StashApplyArgs extends StashArgs {
+  readonly method: "apply" | "restore";
+}
+
+enum LockOrigin {
+  Acquired = 0,
+  NewElement = 1,
+  Discovered = 2,
+}
+
+/**
+ * Stash manager allow stash, drop, apply and merge stashes
+ * @internal
+ */
+export class StashManager {
+
+  private static readonly STASH_ROOT_DIR_NAME: string = ".stashes";
+  /**
+   * Retrieves the root folder path for stash files associated with the specified BriefcaseDb.
+   *
+   * @param db - The BriefcaseDb instance for which to determine the stash root folder.
+   * @param ensureExists - If true, the stash root directory will be created if it does not already exist.
+   * @returns The absolute path to the stash root directory.
+   * @throws IModelError if the database is not open, is readonly, has unsaved changes, or if the briefcase path cannot be determined.
+   */
+  private static getStashRootFolder(db: BriefcaseDb, ensureExists: boolean): LocalDirName {
+    if (!db.isOpen || db.isReadonly)
+      throw new StashError("Database is not open or is readonly");
+
+    if (!existsSync(db[_nativeDb].getFilePath())) {
+      throw new StashError("Could not determine briefcase path");
+    }
+
+    const stashDir = path.join(path.dirname(db[_nativeDb].getFilePath()), StashManager.STASH_ROOT_DIR_NAME, `${db.briefcaseId}`);
+    if (ensureExists && !existsSync(stashDir)) {
+      mkdirSync(stashDir, { recursive: true });
```

Comment:
Should this have an exception handler that throws a `StashError`?

Reply:
If not, the `@throws` part of the comment is probably wrong, since it can also throw whatever error `mkdirSync` throws.

Reply:
fixed. All throw docs are removed.